clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/PartitionConsumptionState.java

```diff
@@ -57,7 +61,41 @@ public class PartitionConsumptionState {
   private boolean completionReported;
   private boolean isSubscribed;
   private boolean isDataRecoveryCompleted;
-  private LeaderFollowerStateType leaderFollowerState;
+
+  /**
+   * The lock is mainly to protect the consumer thread (while processing inflight messages) from the SIT thread trying
+   * to update the state during a state transition. The read lock is acquired in the consumer thread for the entirety
+   * of one message's processing in {@link StoreIngestionTask#produceToStoreBufferServiceOrKafka}, and the SIT must wait
+   * for the release of the read lock before it can acquire the write lock and update the state. That is the only
+   * lengthy acquisition of the lock.
+   */
+  final class LeaderFollowerState {
+    private LeaderFollowerStateType state;
+    private final ReadWriteLock rwLock;
+
+    LeaderFollowerState() {
+      this.state = LeaderFollowerStateType.STANDBY;
+      this.rwLock = new ReentrantReadWriteLock(true); // TODO: would it be better if this was non-fair? I think no
```

Comment:
What does this true mean?

Reply:
The parameter is for whether the lock is in fair-mode or not ([docs link](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html)). Basically, fair-mode will prioritize write locks trying to acquire the lock to ensure that the write lock doesn't get starved. More specifically, it'll respect approximate arrival-order policy whereas non-fair will have no order of entry guarantees.

Reply:
I spoke with @haoxu07 and since the PCS is on the per-partition level, the lock contention should be minimal and only among a few threads. Fair-mode should be unnecessary, and the writer should not be starved.

Reply:
Yeah, as we discussed the L->F/F->L should be very infrequent. The doc does not mention if this will introduce any other perf overhead though, so I guess it is fine